core/lib/snapshots_applier/src/lib.rs

```diff
@@ -0,0 +1,282 @@
+                Fatal(Error::msg(format!("Unable to deserialize object: {err}")))
+            }
+            ObjectStoreError::Other(_) => {
+                Retryable("An error occurred while accessing object store".to_string())
+            }
+        }
+    }
+}
+
+impl From<SqlxError> for SnapshotApplierError {
+    fn from(err: SqlxError) -> Self {
+        match err {
+            SqlxError::Database(_) => Fatal(err.into()),
+            SqlxError::RowNotFound => Fatal(err.into()),
+            SqlxError::ColumnNotFound(_) => Fatal(err.into()),
+            SqlxError::Configuration(_) => Fatal(err.into()),
+            SqlxError::TypeNotFound { type_name: _ } => Fatal(err.into()),
+            _ => Retryable(format!("An error occured when accessing DB: {err}")),
+        }
+    }
+}
+
+#[async_trait]
+pub trait SnapshotsApplierMainNodeClient: fmt::Debug + Send + Sync {
+    async fn miniblock_hash(&self, number: MiniblockNumber) -> Result<Option<H256>, RpcError>;
+
+    async fn fetch_newest_snapshot(&self) -> Result<Option<SnapshotHeader>, RpcError>;
+}
+impl<'a> SnapshotsApplier<'a> {
+    pub async fn load_snapshot(
+        connection_pool: &'a ConnectionPool,
+        main_node_client: Box<dyn SnapshotsApplierMainNodeClient>,
+        blob_store: Box<dyn ObjectStore>,
+    ) -> anyhow::Result<(), SnapshotApplierError> {
+        let mut storage = connection_pool
+            .access_storage_tagged("snapshots_applier")
+            .await?;
+        let mut storage = storage.start_transaction().await?;
+
+        let mut applied_snapshot_status = storage
+            .snapshot_recovery_dal()
+            .get_applied_snapshot_status()
+            .await
+            .map_err(|_| Retryable("Unable to get applied snapshots status".to_string()))?;
+
+        if !storage
+            .blocks_dal()
+            .is_genesis_needed()
+            .await
+            .map_err(|_| Retryable("Unable to fetch genesis state".to_string()))?
+            && applied_snapshot_status.is_none()
+        {
+            return Err(Canceled(
+                "This node has already been initialized without a snapshot".to_string(),
+            ));
+        }
+
+        if let Some(applied_snapshot_status) = applied_snapshot_status.as_ref() {
+            if applied_snapshot_status
+                .storage_logs_chunks_ids_to_process
+                .is_empty()
+            {
+                return Err(Canceled(
+                    "This node has already been initialized from a snapshot".to_string(),
+                ));
+            }
+        } else {
+            applied_snapshot_status =
+                Some(SnapshotsApplier::create_fresh_recovery_status(main_node_client).await?)
+        }
+
+        let mut recovery = Self {
+            connection_pool,
+            blob_store,
+            applied_snapshot_status: applied_snapshot_status.unwrap(),
+        };
+
+        recovery.recover_factory_deps(&mut storage).await?;
```

Comment:
What if factory deps are already recovered?

Reply:
Initial checks and inserting those factory deps are inside a single transaction, so this situation cannot occur

Reply:
I've meant that the same deps may be inserted into Postgres multiple times, but this is fine given that the relevant DB statement has `ON CONFLICT (bytecode_hash) DO NOTHING`.